File: core/src/main/java/y2k/joyreactor/CommentGroup.java
package y2k.joyreactor;

import java.util.List;

/**
 * Created by y2k on 07/12/15.
 */
public interface CommentGroup {

    Comment get(int position);

    int size();

    boolean isChild(int position);

    long getId(int position);

    /**
     * Created by y2k on 11/28/15.
     */
    class OneLevel implements CommentGroup {

        private Comment parent;
        private List<Comment> children;

        public OneLevel(List<Comment> children) {
            this(null, children);
        }

        public OneLevel(Comment parent, List<Comment> children) {
            this.parent = parent;
            this.children = children;
        }

        @Override
        public Comment get(int position) {
            if (parent == null) return children.get(position);
            if (position == 0) return parent;
            return children.get(position - 1);
        }

        @Override
        public int size() {
            return children.size() + (parent == null ? 0 : 1);
        }

        @Override
        public boolean isChild(int position) { // is need divider
            return parent != null && position > 0;
        }

        @Override
        public long getId(int position) {
            if (parent == null) return children.get(position).id;
            if (position == 0) return parent.parentId;
            return children.get(position - 1).id;
        }
    }

    class TwoLevel implements CommentGroup {

        private List<Comment> comments;

        public TwoLevel(List<Comment> comments) {
            this.comments = comments;
        }

        @Override
        public boolean isChild(int position) {
            return comments.get(position).parentId != 0;
        }

        @Override
        public Comment get(int position) {
            return comments.get(position);
        }

        @Override
        public int size() {
            return comments.size();
        }

        @Override
        public long getId(int position) {
            return comments.get(position).id;
        }
    }
}

File: core/src/main/kotlin/y2k/joyreactor/CommentGroup.kt
package y2k.joyreactor

/**
 * Created by y2k on 07/12/15.
 */
interface CommentGroup {

    operator fun get(position: Int): Comment

    fun size(): Int

    fun isChild(position: Int): Boolean

    fun getId(position: Int): Long

    object Empty : CommentGroup {

        override fun size(): Int {
            return 0
        }

        override fun get(position: Int): Comment {
            throw UnsupportedOperationException()
        }

        override fun isChild(position: Int): Boolean {
            throw UnsupportedOperationException()
        }

        override fun getId(position: Int): Long {
            throw UnsupportedOperationException()
        }
    }

    /**
     * Created by y2k on 11/28/15.
     */
    class OneLevel(private val parent: Comment?, private val children: List<Comment>) : CommentGroup {

        constructor(children: List<Comment>) : this(null, children) {
        }

        override operator fun get(position: Int): Comment {
            if (parent == null) return children[position]
            if (position == 0) return parent
            return children[position - 1]
        }

        override fun size(): Int {
            return children.size + if (parent == null) 0 else 1
        }

        override fun isChild(position: Int): Boolean {
            // is need divider
            return parent != null && position > 0
        }

        override fun getId(position: Int): Long {
            if (parent == null) return children[position].id
            if (position == 0) return parent.parentId
            return children[position - 1].id
        }
    }

    class TwoLevel(private val comments: List<Comment>) : CommentGroup {

        override fun isChild(position: Int): Boolean {
            return comments[position].parentId != 0L
        }

        override operator fun get(position: Int): Comment {
            return comments[position]
        }

        override fun size(): Int {
            return comments.size
        }

        override fun getId(position: Int): Long {
            return comments[position].id
        }
    }
}